libnetwork: just forward the external DNS response

corhere opened the pull request:
- Fixes #45553
- Fixes #45555

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

Changes to the DNS resolver in #44664 fixed some bugs, including load-bearing bugs which masked other bugs. One bug not fixed by the aforementioned PR is that the resolver handles non-authoritative NXDOMAIN responses by failing over to the next external DNS server in the list. That behaviour does not make sense as authoritative responses only come from the authoritative name servers for a domain, and most of the time none of the servers in the list will be authoritative. What would make sense for handling a non-authoritative NXDOMAIN is to recursively resolve up to the authoritative name servers to get an authoritative answer to cache. But our DNS resolver is [a stub resolver, not a recursive resolver,](https://datatracker.ietf.org/doc/html/rfc8499#section-6) so there is no good reason for it to treat any NXDOMAIN response differently from a NOERROR response. (If you depended on this fail-over behaviour to successfully resolve domains because the first external DNS server in your list returns NXDOMAIN for the public internet, that DNS server is broken and needs to be fixed. If we can make a split-horizon DNS resolver which resolves public domain names by forwarding, so can you.)

Commit 9cf8c4f68955457f15288701eb59da045349b214 fixed a bug in the fail-over behaviour by replying SERVFAIL if none of the external name servers provided a successful response instead of falling through to forward the last non-successful response received. Unfortunately that was a load-bearing bug which masked the aforementioned NXDOMAIN issue.

**- What I did**
Our resolver is just a forwarder for external DNS—a stub resolver—so it should act like it. Unless it's a server failure or refusal, take the response at face value and forward it along to the client. RFC 8020 is only applicable to caching recursive name servers and our resolver is neither caching nor recursive.

**- How I did it**
Deleted code which had no business existing in a DNS stub resolver.

**- How to verify it**
```console
$ docker network create mybridge
$ docker run --rm -t --network mybridge --dns 8.8.8.8 alpine sh -c 'apk add -qU bind-tools && dig nxdomain.test'
```

v24.0.0:
```
; <<>> DiG 9.18.14 <<>> nxdomain.test
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 45178
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;nxdomain.test.			IN	A

;; Query time: 4 msec
;; SERVER: 127.0.0.11#53(127.0.0.11) (UDP)
;; WHEN: Thu May 18 14:12:18 UTC 2023
;; MSG SIZE  rcvd: 31
```

This PR:
```
; <<>> DiG 9.18.14 <<>> nxdomain.test
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 34098
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;nxdomain.test.			IN	A

;; AUTHORITY SECTION:
.			86398	IN	SOA	a.root-servers.net. nstld.verisign-grs.com. 2023050900 1800 900 604800 86400

;; Query time: 6 msec
;; SERVER: 127.0.0.11#53(127.0.0.11) (UDP)
;; WHEN: Thu May 18 14:15:05 UTC 2023
;; MSG SIZE  rcvd: 117
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
- Fixed an issue where DNS query NXDOMAIN replies from external servers were forwarded to the client as SERVFAIL

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented on libnetwork/resolver.go:
```diff
@@ -479,19 +479,6 @@ func (r *Resolver) forwardExtDNS(proto string, query *dns.Msg) *dns.Msg {
 			// Server returned REFUSED: this can be a transitional status, so continue with the next external DNS server
 			logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), queryName)
 			continue
-		case dns.RcodeNameError:
-			// Server returned NXDOMAIN. Stop resolution if it's an authoritative answer (see RFC 8020: https://tools.ietf.org/html/rfc8020#section-2)
-			logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), queryName)
-			if resp.Authoritative {
-				break
-			}
-			continue
```
Still somewhat on the fence on this, as this would be a breaking change (even if the "split horizon" scenario is wrong?). 

TBH part of https://datatracker.ietf.org/doc/html/rfc1035#page-27 is a bit ambiguous (or at least, I have no idea how it should be interpreted) (what does "meaningful only" mean? Should be ignored otherwise?);

```
3               Name Error - Meaningful only for
                responses from an authoritative name
                server, this code signifies that the
                domain name referenced in the query does
                not exist.
```

corhere replied:
As far as I can tell, "meaningful" is in reference to recursive DNS resolvers, as in a recursive resolver should treat non-authoritative NXDOMAIN responses as a signal to continue recursing to the authoritative name server to get an authoritative response. We are a stub resolver, not a recursive resolver, so our duty is to faithfully return the response provided by the recursive resolver we delegated the handling of the query onto.

corhere commented on libnetwork/resolver.go:
```diff
@@ -479,19 +479,6 @@ func (r *Resolver) forwardExtDNS(proto string, query *dns.Msg) *dns.Msg {
-			continue
-		case dns.RcodeSuccess:
-			// All is well
-		default:
-			// Server gave some error. Log the error, and continue with the next external DNS server
-			logrus.Debugf("[resolver] external DNS %s:%s responded with %s (code %d) for %q", proto, extDNS.IPStr, statusString(resp.Rcode), resp.Rcode, queryName)
-			continue
```
> Should we still log this?

I see no value in logging responses which are forwarded verbatim to the client.

> Do we know what other errors we could receive?

[Yes.](https://pkg.go.dev/github.com/miekg/dns#RcodeSuccess)

> Would it still make sense to fall back to the next DNS, in case the first had an issue?

No, because the only response codes indicating a server issue are already handled by falling back to the next server. Doing the same for any other response code would be like an HTTP reverse proxy falling back on the next server because the first replied with a 4xx response.

thaJeztah replied:
> I see no value in logging responses which are forwarded verbatim to the client.

These logs for sure have helped debugging in the past; they get handled inside the container by `<whatever>` runs inside the container, and having logs on the daemon side could help debugging "hey, it's an issue with your DNS"

thaJeztah replied:
>> Do we know what other errors we could receive?
>
> [Yes.](https://pkg.go.dev/github.com/miekg/dns#RcodeSuccess)

>> Would it still make sense to fall back to the next DNS, in case the first had an issue?
>
> No, because the only response codes indicating a server issue are already handled by falling back to the next server. Doing the same for any other response code would be like an HTTP reverse proxy falling back on the next server because the first replied with a 4xx response.

I guess my question here is are those 4xx responses equivalent for all configured DNS servers? e.g. looking at `VERSION`, `BADALG`, `BADTIME`, and without fully digging through them, some of those (at a glance) seem like "this DNS does not implement X" (but the next DNS that's configured _may_ ?)

thaJeztah replied:
> Unless we dump the entire query and response in the logs, comparing dig vs dig @upstream.dns.server from inside the container would be much more informative for troubleshooting DNS issues.

Reality is that this will be a lot more involved. The usual report will be "docker's DNS doesn't resolve X".

Getting the user to provide daemon logs, and those logs showing "we tried your DNS, and your DNS gave a XX error" means we can show it's (most likely) not an issue with our embedded DNS, and that the user needs to go look elsewhere.

corhere replied:
> I guess my question here is are those 4xx responses equivalent for all configured DNS servers?

Yes, for the same reason as the "split horizon" misconfiguration. All configured upstream servers should be interchangeable as [failover is only for handling unresponsive servers or server failure.](https://datatracker.ietf.org/doc/html/rfc1034#autoid-41)

> The usual report will be "docker's DNS doesn't resolve X".
> 
> Getting the user to provide daemon logs, and those logs showing "we tried your DNS, and your DNS gave a XX error" means we can show it's (most likely) not an issue with our embedded DNS, and that the user needs to go look elsewhere.

And those users aren't going to be running the daemon in debug mode, so they will have to actively do things to change that and collect the logs. What's the difference if we instead instruct them to run `dig` a couple of times?

thaJeztah replied:
> And those users aren't going to be running the daemon in debug mode, so they will have to actively do things to change that and collect the logs. What's the difference if we instead instruct them to run dig a couple of times?

- Minimal images without dig
- Alpine images doing different resolution than non-alpine
- "In my `<insert complex setup>`, things fail"

Whereas (and perhaps some of these logs should not be `Debug()` if they _should be_ infrequent) having a logs that proves "it's not our problem" can avoid all of that, and can save us a lot of time. Daemon logs are also collected in Docker Desktop support dumps, so can provide some of this information after-the-fact.

corhere replied:
https://github.com/moby/moby/blob/de9f12ea962608bd257a07af72a792eba2a38132/libnetwork/resolver.go#L512-L514
We already have a debug log for every reply with no answer section. Is that not sufficient?

corhere committed:
libnetwork: just forward the external DNS response

Our resolver is just a forwarder for external DNS so it should act like
it. Unless it's a server failure or refusal, take the response at face
value and forward it along to the client. RFC 8020 is only applicable to
caching recursive name servers and our resolver is neither caching nor
recursive.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah reviewed (approved):
LGTM

tianon commented:
```
libnetwork/resolver.go:548:6: func `statusString` is unused (unused)
func statusString(responseCode int) string {
     ^
```

seems legit :eyes:

tianon commented:
```
=== Failed
=== FAIL: libnetwork TestProxyNXDOMAIN (0.00s)
    resolver_test.go:407: time="2023-05-18T19:04:10Z" level=debug msg="[resolver] forwarding query" client-addr="udp:127.0.0.1:40278" dns-server="udp:127.0.0.1:58394" question=";example.net.\tIN\t A"
    resolver_test.go:407: time="2023-05-18T19:04:10Z" level=error msg="[resolver] failed to query DNS server: 127.0.0.1:58394, query: ;example.net.\tIN\t A" error="read udp 127.0.0.1:40278->127.0.0.1:58394: read: connection refused"
    resolver_test.go:509: Response:
        ;; opcode: QUERY, status: SERVFAIL, id: 45164
        ;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0
        
        ;; QUESTION SECTION:
        ;example.net.	IN	 A
        
    resolver_test.go:510: Expected DNS response code: 3. Found: 2
```

is this a fluke, or related?

neersighted commented:
Needs cherry-pick to 24

thaJeztah commented:
Just noticed these on Windows unit-tests; perhaps it's part of the test to fail, but it looked .. iffy
https://github.com/moby/moby/actions/runs/5017800744/jobs/8997519683?pr=45565

```
=== RUN   TestCheckoutGit/master:subdir
    gitutils_test.go:211: git-http-backend: fatal: The remote end hung up unexpectedly
        
    gitutils_test.go:211: git-http-backend: 
    gitutils_test.go:211: git-http-backend: fatal: The remote end hung up unexpectedly
        
    gitutils_test.go:211: git-http-backend: 
--- PASS: TestCheckoutGit/master:subdir (5.70s)
```

If those failures are intentional,  we should look at suppressing that output, because it's .. confusing

thaJeztah commented:
Failure is a known flaky;

```
=== FAIL: github.com/docker/docker/libnetwork/networkdb TestNetworkDBCRUDTableEntries (8.23s)
    networkdb_test.go:309: Entry existence verification test failed for node2(647c3a6d1aea)
2023/05/18 21:26:48 Closing DB instances...
```
